pgmpy/inference/mplp.py

```diff
@@ -0,0 +1,302 @@
+        for factor in model.get_factors():
+            scope = frozenset(factor.scope())
+            self.objective[scope] = factor
+            # For every factor consisting of more that a single node, we initialize a cluster.
+            if len(scope) > 1:
+                self.cluster_set[scope] = self.Cluster(self.intersection_set_variables, factor)
+
+        # dual_lp(\delta) is the dual linear program
+        self.dual_lp = sum([max(self.objective[obj].values) for obj in self.objective])
+
+        # Best integral value of the primal objective is stored here
+        self.best_int_objective = 0
+
+        # Assignment of the nodes that results in the "maximum" integral value of the primal objective
+        self.best_assignment = {}
+        # Results of the "maximum" integral value of the primal objective.
+        self.best_decoded_result = {}
+
+    class Cluster:
+
+        """
+        Inner class for representing a cluster.
+        A cluster is a subset of variables.
+
+        Parameters
+        ----------
+        set_of_variables: tuple
+                          This is the set of variables that form the cluster.
+
+        intersection_set_variables: set containing frozensets.
+                                    collection of intersection of all pairs of cluster variables.
+                        For eg: \{\{C_1 \cap C_2\}, \{C_2 \cap C_3\}, \{C_3 \cap C_1\} \} for clusters C_1, C_2 & C_3.
+
+        cluster_potential: Factor
+                           Each cluster has a initial probability distribution provided beforehand.
+        """
+        def __init__(self, intersection_set_variables, cluster_potential):
+            """
+            Initialization of the current cluster
+            """
+
+            # The variables with which the cluster is made of.
+            self.cluster_variables = frozenset(cluster_potential.scope())
+
+            # The cluster potentials must be specified before only.
+            self.cluster_potential = copy.deepcopy(cluster_potential)
+
+            # Generate intersection sets for this cluster; S(c)
+            self.intersection_sets_for_cluster_c = [intersect.intersection(self.cluster_variables)
+                                                    for intersect in intersection_set_variables
+                                                    if intersect.intersection(self.cluster_variables)]
+
+            # Initialize messages from this cluster to its respective intersection sets
+            # \lambda_{c \rightarrow \s} = 0
+            self.message_from_cluster = {}
+            for intersection in self.intersection_sets_for_cluster_c:
+                # Present variable. It can be a node or an edge too. (that is ['A'] or ['A', 'C'] too)
+                present_variables = list(intersection)
+
+                # Present variables cardinality
+                present_variables_card = [cluster_potential.cardinality[cluster_potential.scope().index(variable)]
```

Comment:
Don't do such complex operation to just get the cardinality of a variable. Instead you can simply use

```python
present_variables_card = [cluster_potential.get_cardinality(variable) 
                          for variable in present_variables]
```